### PR TITLE
Use `std::ops` traits for `Exprs` rather than custom function names

### LINKED ISSUE #6437: Inconsistencies in `Expr` operator code.
# Problem

Some operations are implemented both directly on `Expr` and as built-in traits (here's an example: [trait impl](https://github.com/apache/arrow-datafusion/blob/32144501be634253b254aca34660b4b3e81dbc1f/datafusion/expr/src/operator.rs#L299-L305C2) compared with [type impl](https://github.com/apache/arrow-datafusion/blob/32144501be634253b254aca34660b4b3e81dbc1f/datafusion/expr/src/expr.rs#LL744C1-L746C6)).

As far as I can tell, these implementations could (and should) use the same underlying logic (the trait impls should use `Expr`'s methods). Instead, the logic is duplicated at both sites (as you can see from the snippets I linked). This increases the potential for bugs as it would be very easy to introduce changes in say one of `Expr`'s methods and not the corresponding trait impl (or vice-versa).

Another inconsistency is that `Expr`'s methods don't have the same name as the corresponding trait methods, so we end up in a situation where, for example, `bitwise_and` (type impl), `bitand` and `&` (trait impl) all do the same thing. This can cause confusion for users since it's not immediately clear that these are in fact intended to be the same.

Finally, the `Not` trait implementation is located in the `expr` module ([here](https://github.com/apache/arrow-datafusion/blob/32144501be634253b254aca34660b4b3e81dbc1f/datafusion/expr/src/expr.rs#L912-L938)), while all other operator trait impls are located in the `operator` module.

# Potential solution
In my opinion, the `Expr` impl should not duplicate built-in operator trait methods. The duplicated methods should be removed from `Expr`, and only the trait impls be kept. This is more in line with idiomatic Rust code, removes duplication, and reduces potential user confusion.

Likewise, the `Not` trait impl should be moved to the `operator` module and the corresponding type method be removed.

# Alternatives
If the duplicated type impls won't be removed, then the trait impls should be derived from the `Expr` methods to at least remove the duplication in implementation.

# Discussion
Please let me know what you think. It's possible that I misunderstood why the code is laid out this way.

If there is in fact a problem, then I'm prepared to make a PR to solve this after we've discussed of the appropriate solution.

### PR DESCRIPTION
**Pull request by LouisGariepy**
# Which issue does this PR close?

Closes #6437.

# Rationale for this change

Some operators on `Expr` are duplicated between type methods and trait impls. This PR removes the duplicated operators in favor of the trait impls.

See #6437 for a more in-depth explanation of why this should be fixed.

# What changes are included in this PR?

Moved the `Not` trait impl to the `operator` module.

Removed the duplicate type methods.

Fixed calls to the removed methods to use the corresponding trait method in instances where the two method names were not the same.

# Are these changes tested?

No features were added or removed, so the existing tests should be OK.

I removed a now useless test case for the `Not` trait impl, and added a new one in the `operator` module's unit tests. The new test was made in the same style as the other existing operator unit tests.

# Are there any user-facing changes?

Yes, some operator methods have now changed names (when the type methods and the corresponding trait method did not have the same name), and they will now require the relevant trait to be imported.

### CODE REVIEW

**Comment by alamb** [comment]
Thank you @LouisGariepy  -- I plan to review this tomorrow or Tuesday

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 1]
```diff
@@ -1176,7 +1178,11 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
 
 #[cfg(test)]
 mod tests {
-    use std::{collections::HashMap, sync::Arc};
+    use std::{
+        collections::HashMap,
+        ops::{BitAnd, BitOr, BitXor},
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1183]: 💯  for using standard traits

**Review by alamb — APPROVED** [review]
Makes sense to me. Thank you @LouisGariepy 

@izveigor  do you have any concern about this PR?

**Review by jackwener — APPROVED** [review]
Nice cleanup, thank you @LouisGariepy

**Comment by alamb** [comment]
Thanks @LouisGariepy  -- I took the liberty of merging up from main to resolve a conflict on this branch:

Screenshot 2023-05-30 at 9 29 39 AM

I plan to merge when the checks have completed

**Pull request merged**